src/Shared/PackageManagers/PackageManagerFactory.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CST.OpenSource.Shared
+{
+    public class ProjectManagerFactory
+    {
+        // do reflection only once
+        static List<Type> projectManagers = new List<Type>();
+
+        /// <summary>
+        /// Get the project manager for the package type
+        /// </summary>
+        /// <param name="purl"></param>
+        /// <returns>BaseProjectManager object</returns>
+        public static BaseProjectManager CreateProjectManager(PackageURL purl, string destinationDirectory)
+        {
+            if(projectManagers.Count == 0)
+            {
+                projectManagers.AddRange(typeof(BaseProjectManager).Assembly.GetTypes()
+               .Where(type => type.IsSubclassOf(typeof(BaseProjectManager))));
+            }
+            // Use reflection to find the correct package management class
+            var downloaderClass = projectManagers
+               .Where(type => type.Name.Equals($"{purl.Type}ProjectManager",
+                                               StringComparison.InvariantCultureIgnoreCase))
+               .FirstOrDefault();
+            if (downloaderClass != null)
+            {
+                var ctor = downloaderClass.GetConstructor(Array.Empty<Type>());
+                var _downloader = (BaseProjectManager)(ctor.Invoke(Array.Empty<object>()));
+
+                // TODO: find a better way to do this, preferably as constructor argument
+                _downloader.TopLevelExtractionDirectory = destinationDirectory;
```

Comment:
You could use pattern matching and switch here instead of reflection.

```csharp
if (downloaderClass != null){
    switch(downloaderClass){
        case ClassTypeOne:
            return new ClassTypeOne(destinationDirectory);
        ...
    }
}
```

Reply:
Rationale for reflection here was to not have to hardcode each of the 10+ different subclasses. I'd prefer a more elegant way of representing this if possible, though.

Reply:
I don't mind doing reflection here, especially with this refactor, since we are doing it only once. If performance is not extremely critical, its the easiest (and only) way to get the derived classes dynamically. We can revisit it if and when performance is a bottleneck. 
Statically doing if else requires modifying code each time a new class is added.

Reply:
Good point. I do like that won't have to rewrite this. 

I'd rewrite all the constructors to take that argument though.

Reply:
Agreed on that.